File: 019RemoveNthNodeFromEndOfList.cpp
//1. Keep two pointers. One is n step forward than the others. When this one reach the end, another is at the nth. O(N)
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* cur, *pre;
        cur = head; pre = head;
        
        int i=n;
        while(i-->0) pre = pre->next;  //Question assume that the n<list.length()
        if(!pre) return head->next;
        
        while(pre->next){ //This is important. At the end pre is at last node, cur is at last (n+1) node.
            pre = pre->next;
            cur = cur->next;
        }
        cur -> next = cur -> next -> next;
        
        return head;
    }
};

File: Algo/Leetcode/019RemoveNthNodeFromEndOfList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        move = head
        fast = head
        while fast and n > 0:
            fast = fast.next
            n -= 1

        while fast:
            fast = fast.next
            move = move.next

        # first
        if head == move:
            return head.next

        # remove
        if move.next is None:  # last
            fast = head
            while fast.next.next:
                fast = fast.next
            fast.next = None
        else:   # not last
            move.val = move.next.val
            move.next = move.next.next

        return head
